python/setup.py

```diff
@@ -22,15 +22,31 @@
 import sys
 # need to use distutils.core for correct placement of cython dll
 kwargs = {}
+required_packages = ['numpy', 'requests', 'graphviz']
 if "--inplace" in sys.argv:
     from distutils.core import setup
     from distutils.extension import Extension
+    from distutils.command.install import install
 else:
     from setuptools import setup
     from setuptools.extension import Extension
-    kwargs = {'install_requires': ['numpy', 'requests', 'graphviz'], 'zip_safe': False}
+    kwargs = {'install_requires': required_packages,
+              'setup_requires': required_packages,
+              'tests_require': required_packages,
+              'zip_safe': False}
+    from setuptools.command.install import install
 from setuptools import find_packages
 
+class PostInstallCommand(install):
+    """Post-installation for installation mode."""
+    def run(self):
+        install.run(self)
+        from mxnet.base import _generate_op_module_signature
```

Comment:
Is this guaranteed to import the mxnet that's just installed? Can you read the path from install or self?

Reply:
There's no guarantee, and neither is there guarantee that self or install resides in the target site-packages. Currently what this code does in practice is:
1. generate the code in the source code folder first (hence the .gitignore change)
2. copy source codes into $PWD/build/mxnet and add packaging info
3. copy $PWD/build to the site-packages and byte-compile the .py files.

Reply:
What happens for pip package?

Reply:
It copies the whole tree from `mxnet-distro/mxnet-build/python/mxnet` (without generated source) to `mxnet-distro/mxnet`, generates the source then and there, and does the above step 2. Finally, the whole tree in `mxnet-distro/build` is packaged into a wheel which contains the generated source.